### DISCUSSION TITLE
Terraform provisioning always going to network boot

### DISCUSSION THREAD

**Discussion by thatsk** [terraform-coop/terraform-provider-foreman#105] [Q&A]
Hello Team, I am trying to use image as provisioned but by default it is going to install via network boot. Is there any param we need to set to use for cloning VM template.

**Comment by thatsk** [comment]
@lhw any lights on this

**Comment by lhw** [comment]
Something like this should work:

```hcl
data "foreman_image" "example" {
  name = "some image name"
}
resource "foreman_host" "example" {
  name = "compute01.dc1.company.com"
  image_id = data.foreman_image.example.id
}

**Comment by thatsk** [comment]
this is error @lhw 
```
 "error": {"id":null,"errors":{"ptable_id":["can't be blank unless a custom partition has been defined"],"medium_id":["can't be blank"]},"full_messages":["Ptable can't be blank unless a custom partition has been defined","Medium can't be blank"]}
}
]
}
```

**Comment by thatsk** [comment]
This was working upto 0.4.5 and now latest seems like some new stuff has been added.
```

resource "foreman_host" "newtest" {
    name                    = "kunal"
    compute_profile_id      = "${data.foreman_computeprofile.medium.id}"
    build                   = true
    compute_resource_id     = "${data.foreman_computeresource.int-msp.id}"
    domain_id               = 2
    hostgroup_id            = "${data.foreman_hostgroup.terraform.id}"
    image_id                = "${data.foreman_image.example.id}"
    model_id                = 0
    operatingsystem_id  = "${data.foreman_operatingsystem.Centos79.id}"
    owner_id            = 3
    retry_count         = 2
    parameters = {
      "puppet_environment" = "farm",
    }
}
resource "null_resource" "newtest" {
  triggers = {
       templateid = "${foreman_host.newtest.id}"
  }

  connection {
    type = "ssh"
    host = "${foreman_host.newtest.interfaces_attributes[0].ip}"
    user = "root"
    password = "${var.password}"
    port = "22"
    agent = false
  }

  provisioner "file" {
      source      = "scripts/"
      destination = "/tmp"
 }
}
```

**Comment by thatsk** [comment]
when I am checking the code it has been refactored for provision_method and build params.

**Comment by thatsk** [comment]
which param is defining i need to use this image for cloning and which param say it goes in network boot

**Comment by thatsk** [comment]
on 0.5.0 says methods are deprecated now and need to use method and build and if we get some lights on this.
https://github.com/terraform-coop/terraform-provider-foreman/releases/tag/v0.5.0

**Comment by thatsk** [comment]
https://github.com/terraform-coop/terraform-provider-foreman/commit/f39044833265a92c8f47f04075da72db12a879e2#diff-cd3952521384329d0a0e1d6a3520280acb84b99f14aa1382053142c13b029896

@lhw i am trying to understand and findiing solution but this is not working even I use build true or false

**Comment by thatsk** [comment]
it says use build arg how to use that in terraform code

**Comment by thatsk** [comment]
If you can guide me what I am doing wrong then I will knw

**Comment by thatsk** [comment]
@lhw any pointer?

**Comment by agriffit79** [comment]
Is this still an issue? If so can you post DEBUG log output for it working in the older provider version and not working in the latest version. Specifically the line containing `hJSONBytes:`

**Reply by thatsk** [reply]
is there any bug in my above code or is that correct

**Comment by thatsk** [comment]
@agriffit79 @lhw 
Working in 0.4.x 
```
hJSONBytes: [{"host":{"build":false,"comment":"","compute_attributes":{},"compute_profile_id":"160","compute_resource_id":"38","domain_id":"133","host_parameters_attributes":[{"name":"puppet_environment","value":"farm"}],"hostgroup_id":"375","image_id":"12","medium_id":null,"model_id":null,"name":"kunal","operatingsystem_id":"89","owner_id":"3","owner_type":"","provision_method":"image","puppetclass_ids":[]},"location_id":1,"organization_id":2}]
```

Not working In 0.5.x which has introduced build parameter:-
````
hJSONBytes: [{"host":{"id":0,"name":"kunal","created_at":"","updated_at":"","build":false,"domain_id":133,"owner_id":3,"hostgroup_id":375,"operatingsystem_id":89,"image_id":12,"managed":true,"comment":"","host_parameters_attributes":[{"name":"puppet_environment","value":"farm"}],"compute_resource_id":38,"compute_profile_id":160,"config_group_ids":null,"puppet_attributes":{"puppetclass_ids":null,"config_group_ids":null}},"location_id":1,"organization_id":2}]
```

**Reply by thatsk** [reply]
provision method is missing in new one

**Reply by thatsk** [reply]
how its handled in build parameter thats confusing part

**Comment by thatsk** [comment]
https://github.com/terraform-coop/terraform-provider-foreman/issues/106

**Comment by thatsk** [comment]
waiting for your response

**Comment by agriffit79** [comment]
This is another regression caused by 2fa9ff2. It's now impossible to set `provision_method = image`. I think the best fix is to set this whenever `image_id` is a non nil value.

Are you able to locally build and test the provider if I do a quick hack to test it will work?

**Reply by thatsk** [reply]
what should i test can you point me ? i would happy to test it.

**Reply by thatsk** [reply]
yes if you can point me branch with new version or some commit I would happy to test it.

**Reply by agriffit79** [reply]
Clone this repo and build from the `fix_clone` branch https://github.com/agriffit79/terraform-provider-foreman.git

**Reply by thatsk** [reply]
k doing it now

**Reply by thatsk** [reply]
@agriffit79 
~/terraform-provider-foreman$ /usr/local/bin/goreleaser build --skip-validate
  • starting build...
  • loading config file                              file=.goreleaser.yml
  • loading environment variables
  • getting and validating git state
    • building...                                    commit=e515ffe9559e89ff93d6dc7c12fc17e4b281498f latest tag=v0.4.3
    • pipe skipped                                   reason=validation is disabled
  • parsing tag
  • setting defaults
  • running before hooks
    • running                                        hook=go mod tidy
  • checking distribution directory
  • loading go mod information
  • build prerequisites
  • writing effective config file
    • writing                                        config=dist/config.yaml
  • building binaries
    • building                                       binary=dist/terraform-provider-foreman_linux_amd64_v1/terraform-provider-foreman_v0.4.3
    • took: 2s
  • storing release metadata
    • writing                                        file=dist/artifacts.json
    • writing                                        file=dist/metadata.json
  • build succeeded after 2s

**Reply by thatsk** [reply]
its building for 0.4.3 is that ok with 0.5 ? does it gonna work with above example which is on 0.5.x

**Reply by thatsk** [reply]
@agriffit79 tested cloning works fine can you raise PR against 0.5.8 so that I can test in official release?

**Reply by agriffit79** [reply]
I would like to see if we can get a fix for #106 as well before releasing a new version. To avoid too much churn in the code base.

**Reply by thatsk** [reply]
sounds good thank you @agriffit79 much appreciated

**Comment by thatsk** [comment]
```
resource "foreman_host" "host1" {
    name                    = "kunal"
    compute_profile_id      = "${data.foreman_computeprofile.medium.id}"
    compute_resource_id     = "${data.foreman_computeresource.farm1.id}"
    build                   = true
    domain_id               = 133
    hostgroup_id            = "${data.foreman_hostgroup.farm1.id}"
    image_id                = "${data.foreman_image.centos7.id}"
    model_id                = 0
    operatingsystem_id  = "${data.foreman_operatingsystem.Centos79.id}"
    owner_id            = 3 
    retry_count         = 2
    parameters = {
      "puppet_environment" = "farm",
    }
}
```

**Comment by thatsk** [comment]
@agriffit79  its not cloning the vm can you tell me if I missed anything? this is not working in 0.5.x

**Comment by thatsk** [comment]
https://github.com/terraform-coop/terraform-provider-foreman/issues/103

**Comment by thatsk** [comment]
https://github.com/terraform-coop/terraform-provider-foreman/issues/102

**Comment by thatsk** [comment]
@agriffit79 don't why owner has closed this issue?

**Comment by bitkeks** [comment]
Problem was, the "method" (now called "provision_method") argument was not passed on to Foreman, therefore it was always a "build". See PR #122 which re-introduces the method and works with image-based provisioning.
